x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/RemoteAccessAuthentication.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptorsIntersection;
+
+import java.io.IOException;
+import java.util.Base64;
+
+public record RemoteAccessAuthentication(Authentication authentication, RoleDescriptorsIntersection roleDescriptorsIntersection) {
+
+    private static final String REMOTE_ACCESS_AUTHENTICATION_HEADER = "_remote_access_authentication";
+
+    public String encode() throws IOException {
+        BytesStreamOutput output = new BytesStreamOutput();
+        output.setVersion(authentication.getEffectiveSubject().getVersion());
+        Version.writeVersion(authentication.getEffectiveSubject().getVersion(), output);
+        writeTo(output);
+        return Base64.getEncoder().encodeToString(BytesReference.toBytes(output.bytes()));
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        authentication.writeTo(out);
+        roleDescriptorsIntersection.writeTo(out);
```

Comment:
We might want to wrap `roleDescriptorsIntersection` as raw bytes so it can be deserialised as raw bytes without having to build the object. The raw bytes are useful for caching purpose similar to how API key role descriptor bytes are used. My current thinking is that we can have a new subject type on FC which returns a new roleReference. This new roleReference has a field that is the raw bytes and it knows how to compute the cacheKey (via the `id()` method) and resolve the roleDescriptors (by deserialising the raw bytes into the actual RoleDescriptor object).

As for the raw bytes, it can be either (1) a base64 encoding of `roleDescriptorsIntersection.writeTo(out)` or (2) `roleDescriptorsIntersection.toXContent`.

Reply:
Btw, my suggestion of using raw bytes is mainly from a performance point of view. Otherwise, it is functionally possible to always work with the actual `RoleDescriptor` object. It can still compute a similar cache key with overhead of extra de/serialisations. 

If we are going to raw bytes, for caching purpose, we might even want to consider serialising each role descriptor separately and skip the `RoleDescriptorIntersection` container. Since the role descriptors are built from `Role`, we should always have a simpler `List<RoleDescriptor>` instead of `List<Set<RoleDescriptor>>` as mandated by `RoleDescriptorIntersection`. Having each role descriptor separately is better for more effective caching. 

Please let me know if this makes sense to you. Or maybe it is an implementation detail out scope of this draft PR? Happy to discuss.

Reply:
Discussed on Slack. 

It's indeed redundant work to deserialize the header into `RoleDescriptorIntersection` only to serialize it again.

However, sending raw bytes for `RoleDescriptorIntersection` does not work well with how we construct roles and caching. I think the most sensible approach is to send a `Collection<ByteReference> roleDescriptorsBytesIntersection` where the inner `ByteReference` is an XContent-serialized `Set<RoleDescriptor>` instance. This plays well with how `RoleReference` works. I've pushed this change now. 

Great suggestion, as usual!